### PR TITLE
Refactoring mattermost package, to suit for distribution 

### PR DESCRIPTION
**Pull request by phoinixgrr**
#### Summary

We are introducing the following essential changes for `matermost` package, to better suit for distribution: 

- Change binary destination to `/usr/bin/` instead of `/usr/local/bin/` , to align with best practises. 
 `/usr/bin` is for distribution-managed normal user programs.
  `/usr/local/bin` is for normal user programs not managed by the distribution package manager, e.g. locally compiled packages. 
- Renaming config file from `config.json` to `config.defaults.json` to keep in case of upgrade/purge remove 
- Introduce logic, to define if this is a first time installation, or an upgrade 
- Create mattermost OS user/grp and define relevant permissions 
- Remove mattermost OS user/grp in case of removal
- Perform file cleanup(old mm version) in case of upgrade 
- Introduce env variable:`MM_INSTALL_TYPE` under runtime
- Remove postgresql dependency for `mattermost.service`

#### Ticket Link

**Ticket:** https://mattermost.atlassian.net/browse/CLD-4311

### CODE REVIEW

File: Makefile
Comment (spirosoik) [inline Makefile:52]: won't this change affect omnibus? I dunno that's why I am asking

File: mattermost/DEBIAN/postinst
Comment (spirosoik) [inline mattermost/DEBIAN/postinst:9]: I think you are trying to do this? 
`MATTERMOST_USER=${MATTERMOST_USER:-mattermost}`

same for group

File: mattermost/DEBIAN/postinst
Comment (spirosoik) [inline mattermost/DEBIAN/postinst:4]: nit: you can do something like

```
TRUE_REG='^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$'
FALSE_REG='^([fF][aA][lL][sS][eE]|[nN]|[nN][oO]|0)$'
```

and then do:
```
if [[ $IS_UPGRADE =~ $TRUE_REG ]]; then
```

Also is a good practice to add some debug flag and strict scripting to help us debug when is needed:
```
IS_DEBUG=${IS_DEBUG:-false}
if [[ $IS_DEBUG =~ $TRUE_REG ]]; then
    set -o xtrace
fi

IS_STRICT=${IS_STRICT:-false}
if [[ $IS_STRICT =~ $TRUE_REG ]]; then
    set -o errexit
    set -o nounset
    set -o pipefail
fi
```

File: mattermost/DEBIAN/postrm
Comment (spirosoik) [inline mattermost/DEBIAN/postrm:8]: same as I mentioned above

File: mattermost/files/mattermost.service
Comment (spirosoik) [inline mattermost/files/mattermost.service:15]: is this the right env variable? the env value needs to change to the correct value right?

**Review by spirosoik — CHANGES_REQUESTED** [review]
few minor changes and recommendations. Nothing major

File: Makefile
Comment (phoinixgrr) [reply Makefile:52]: Nope. Omnibus is tested and works as expected. 
mattermost binary symbolic link is still available on the OS

File: mattermost/DEBIAN/postinst
Comment (phoinixgrr) [reply mattermost/DEBIAN/postinst:9]: We are trying to perform the following: 
If `MATTERMOST_USER` is null (not set) we set ot to `mattermost` 
Same for `MATTERMOST_GROUP`

File: mattermost/files/mattermost.service
Comment (phoinixgrr) [reply mattermost/files/mattermost.service:4]: We remove `postgresql` hardcoded dependency from mattermost-server packaging, since it is in the user's repsonsibility/choice to select his favorite db service (postgresql - mysql) 
Keeping it non opinionated. 

When shipping with omnibus-mattermost, we define this as postgresql (Highly opinionated)

File: mattermost/files/mattermost.service
Comment (phoinixgrr) [reply mattermost/files/mattermost.service:15]: Correct. Waiting for some feedback @ https://docs.google.com/document/d/1AXr0aip1pC1p62K-abDGX8rdWv9MtX9UiT8TLaz2iUY/edit?disco=AAAAh9_JbUg to finilize this. 

The functionality is implemented

File: mattermost/DEBIAN/postinst
Comment (phoinixgrr) [reply mattermost/DEBIAN/postinst:4]: These are getting shipped with the mattermost debian packages. Are we sure we want to introduce debugging flags to our released packages? 

Example 
https://github.com/grafana/grafana/blob/main/packaging/deb/control/postinst

File: mattermost/DEBIAN/postrm
Comment (phoinixgrr) [reply mattermost/DEBIAN/postrm:8]: We are trying to perform the following:
If MATTERMOST_USER is null (not set) we set to to `mattermost`
Same for MATTERMOST_GROUP

File: mattermost/DEBIAN/postinst
Comment (spirosoik) [reply mattermost/DEBIAN/postinst:9]: yeah this is my intent too with what I shared. you can give a try

File: mattermost/DEBIAN/postrm
Comment (spirosoik) [reply mattermost/DEBIAN/postrm:8]: same as above

File: mattermost/DEBIAN/postinst
Comment (phoinixgrr) [reply mattermost/DEBIAN/postinst:9]: Tested it, seems to work as expected. 

What do we gain with the proposal? 
Wouldn't you agree that it is a more readable form? :
`[ -z "$MATTERMOST_USER" ] && MATTERMOST_USER="mattermost"` 
vs 
`MATTERMOST_USER=${MATTERMOST_USER:-mattermost}` 

The latter requires some bash insights.  

Would vote in favor of readability against saving a couple of chars. 

Let me know what do you think

File: mattermost/DEBIAN/postrm
Comment (phoinixgrr) [reply mattermost/DEBIAN/postrm:8]: Tested it, seems to work as expected.

What do we gain with the proposal?
Wouldn't you agree that it is a more readable form? :
[ -z "$MATTERMOST_USER" ] && MATTERMOST_USER="mattermost"
vs
MATTERMOST_USER=${MATTERMOST_USER:-mattermost}

The latter requires some bash insights.

Would vote in favor of readability against saving a couple of chars.

Let me know what do you think

File: mattermost/files/mattermost.service
Comment (spirosoik) [reply mattermost/files/mattermost.service:4]: ok so for omnibus will stay as is got it.

File: mattermost/files/mattermost.service
Comment (spirosoik) [reply mattermost/files/mattermost.service:15]: awesome, thanks for taking care of this 💪

File: mattermost/files/mattermost.service
Comment (phoinixgrr) [reply mattermost/files/mattermost.service:4]: That's correct

File: mattermost/DEBIAN/postinst
Comment (spirosoik) [reply mattermost/DEBIAN/postinst:4]: Yeah I got it but I believe this will be really helpful if a customer mentions it doesn't work and somehow be able to debug and get some details back in which line etc. failed. In an on-prem environment will be super hard to debug. Thoughts?

Comment (phoinixgrr) [reply mattermost/DEBIAN/postinst:4]: Valid point. 👍 
Will push a Commit to support debugging

Comment (phoinixgrr) [reply mattermost/DEBIAN/postinst:5]: Done with: 


```bash
> sudo IS_DEBUG=true dpkg -i mattermost_7.3.1-0.deb
Selecting previously unselected package mattermost.
(Reading database ... 62439 files and directories currently installed.)
Preparing to unpack mattermost_7.3.1-0.deb ...
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
Unpacking mattermost (7.3.1-0) ...
Setting up mattermost (7.3.1-0) ...
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
+ '[' -z '' ']'
+ MATTERMOST_USER=mattermost
+ '[' -z '' ']'
+ MATTERMOST_GROUP=mattermost
+ getent group mattermost
+ addgroup --system mattermost --quiet
+ id mattermost
+ adduser --system --home /opt/mattermost --no-create-home --ingroup mattermost --disabled-password --shell /bin/false mattermost
+ chown -R mattermost:mattermost /opt/mattermost
+ '[' '' '!=' '' ']'
+ [[ ! false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ systemctl daemon-reload
+ systemctl enable mattermost
> sudo IS_DEBUG=true dpkg -i mattermost_7.4.0-0.deb
(Reading database ... 66288 files and directories currently installed.)
Preparing to unpack mattermost_7.4.0-0.deb ...
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
+ systemctl stop mattermost
+ '[' 7.3.1-0 '!=' '' ']'
+ IS_UPGRADE=true
+ [[ true =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ find /opt/mattermost/ /opt/mattermost/client/ -mindepth 1 -maxdepth 1 '!' '(' -type d '(' -path /opt/mattermost/client -o -path /opt/mattermost/client/plugins -o -path /opt/mattermost/config -o -path /opt/mattermost/logs -o -path /opt/mattermost/plugins -o -path /opt/mattermost/data ')' -prune ')' -exec rm -r '{}' ';'
Unpacking mattermost (7.4.0-0) over (7.3.1-0) ...
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
Setting up mattermost (7.4.0-0) ...
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
+ '[' -z '' ']'
+ MATTERMOST_USER=mattermost
+ '[' -z '' ']'
+ MATTERMOST_GROUP=mattermost
+ getent group mattermost
+ id mattermost
+ chown -R mattermost:mattermost /opt/mattermost
+ '[' 7.3.1-0 '!=' '' ']'
+ IS_UPGRADE=true
+ [[ ! true =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ systemctl daemon-reload
+ systemctl start mattermost
```

Works great 🚀

File: mattermost/DEBIAN/postinst
Comment (spirosoik) [reply mattermost/DEBIAN/postinst:22]: I like the second opinion but 0/5 not strong opinion. minor thing to change

**Review by spirosoik — APPROVED** [review]
LGTM 💪

File: mattermost/DEBIAN/postinst
Comment (stafot) [reply mattermost/DEBIAN/postinst:22]: Alos prefer the 2nd one, but also 0/5

**Review by stafot — DISMISSED** [review]

**Comment by phoinixgrr** [comment]
Applied the variable substitution changes requested 
```
MATTERMOST_USER=${MATTERMOST_USER:-mattermost}
MATTERMOST_GROUP=${MATTERMOST_GROUP:-mattermost}
```

Rebased and dropped your reviews @stafot  @spirosoik  
Thanks for re-reviewing

**Review by stafot — APPROVED** [review]

**Review by spirosoik — APPROVED** [review]
one more time awesome work @phoinixgrr

**Pull request merged**
